### ISSUE TITLE
feat(compat): add XGBoost 3.2 compatibility for custom objectives

### ISSUE THREAD

**Issue by egordm** [OpenSTEF/openstef#865]
## Background

XGBoost 3.2 (Feb 2026) introduced significant changes to multi-target tree support with vector-leaf and changes to how multi-output data flows through the sklearn interface's `_objective_decorator`. This broke OpenSTEF's custom objective functions.

As a **temporary fix**, XGBoost was pinned to `<3.2` in #864. This issue tracks the proper fix.

## Root Cause

The custom objectives (`arctan_loss_multi_objective`, `pinball_loss_multi_objective`) in `packages/openstef-models/src/openstef_models/utils/loss_functions.py` assume XGBoost passes a **1D flattened array** to the objective:

```python
n_items = len(y_true)
n_quantiles = len(quantiles)
n_rows = n_items // n_quantiles          # ← fragile assumption
y_pred = np.reshape(y_pred, (n_rows, n_quantiles))
y_true = np.reshape(y_true, (n_rows, n_quantiles))
```

Similarly in `mean_pinball_loss` (evaluation function):
```python
y_pred = np.reshape(y_pred, [-1, n_quantiles])
```

XGBoost 3.2 may pass **2D `(n_samples, n_quantiles)` arrays** directly to the custom objective when using `multi_strategy="one_output_per_tree"`, making the manual reshape unnecessary or wrong.

Also need to verify that `sample_weight_eval_set` still works correctly in 3.2.

## Scope

Files to investigate and update:
- `packages/openstef-models/src/openstef_models/utils/loss_functions.py` — objective functions (`arctan_loss_multi_objective`, `pinball_loss_multi_objective`, `xgb_prepare_target_for_objective`)
- `packages/openstef-beam/src/openstef_beam/metrics/metrics_probabilistic.py` — `mean_pinball_loss` (used as eval metric)
- `packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py` — `fit()` / `predict()` reshape logic

## Acceptance Criteria

- [ ] A minimal reproduction script that demonstrates the failure with XGBoost 3.2
- [ ] Objective functions handle both 1D (old) and 2D (new) input shapes, or exclusively use the 2D API
- [ ] All existing XGBoost forecaster tests pass with XGBoost 3.2 installed
- [ ] Version constraint in `packages/openstef-models/pyproject.toml` relaxed back to `>=3,<4` (or `>=3.2,<4` if minimum is bumped)
- [ ] `uv.lock` updated

## Notes

- XGBoost 3.2 changelog: https://xgboost.readthedocs.io/en/latest/changes/v3.2.0.html
- Advanced custom objectives guide: https://xgboost.readthedocs.io/en/latest/tutorials/advanced_custom_obj.html
- The `multi_strategy="one_output_per_tree"` param docs: https://xgboost.readthedocs.io/en/latest/tutorials/multioutput.html
- `_objective_decorator` in the sklearn interface wraps the custom fn and passes `dmatrix.get_label()` / weights to it — check what shape this returns in 3.2 for multi-output models
